driver/src/main/com/mongodb/client/gridfs/GridFSDownloadStreamImpl.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright 2015 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.gridfs;
+
+import com.mongodb.MongoGridFSException;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.gridfs.model.GridFSFile;
+import org.bson.BsonBinary;
+import org.bson.BsonDocument;
+import org.bson.BsonInt32;
+import org.bson.BsonValue;
+
+import static com.mongodb.assertions.Assertions.notNull;
+import static java.lang.String.format;
+
+class GridFSDownloadStreamImpl extends GridFSDownloadStream {
+    private final GridFSFile fileInfo;
+    private final MongoCollection<BsonDocument> chunksCollection;
+    private final BsonValue fileId;
+    private final long length;
+    private final int chunkSizeInBytes;
+    private final int numberOfChunks;
+    private int chunkIndex;
+    private int bufferOffset;
+    private long currentPosition;
+    private byte[] buffer = null;
+    private boolean eof;
+    private boolean closed;
+
+    GridFSDownloadStreamImpl(final GridFSFile fileInfo, final MongoCollection<BsonDocument> chunksCollection) {
+        this.fileInfo = notNull("file information", fileInfo);
+        this.chunksCollection = notNull("chunks collection", chunksCollection);
+
+        fileId = fileInfo.getId();
+        length = fileInfo.getLength();
+        chunkSizeInBytes = fileInfo.getChunkSize();
+        numberOfChunks = (int) Math.ceil((double) length / chunkSizeInBytes);
+    }
+
+    @Override
+    public GridFSFile getGridFSFile() {
+        return fileInfo;
+    }
+
+    @Override
+    public int read() {
+        byte[] b = new byte[1];
+        int res = read(b);
+        if (res < 0) {
+            return -1;
+        }
+        return b[0] & 0xFF;
+    }
+
+    @Override
+    public int read(final byte[] b) {
+        return read(b, 0, b.length);
+    }
+
+    @Override
+    public int read(final byte[] b, final int off, final int len) {
+        checkClosed();
+
+        if (eof) {
+            return -1;
+        } if (currentPosition == length) {
+            eof = true;
+            int chunkToCheck = chunkIndex;
+            if (buffer != null) {
+                chunkToCheck += 1;
+            }
+            BsonDocument chunk = getChunk(chunkToCheck);
+            if (chunk != null) {
+                validateData(chunk.getBinary("data", new BsonBinary(new byte[0])).getData(), chunkToCheck);
```

Comment:
What's the second argument to getBinary? A default value?

Don't we want some other kind of error if the "data" field is missing?